bootstrap-app/src/cmr/bootstrap/services/dispatch/core_async_dispatch.clj

```diff
@@ -0,0 +1,143 @@
+  "Bulk index all the tags, acls, and access-groups."
+  [this context start-index]
+  (let [channel (:system-concept-channel this)]
+    (info "Adding bulk index request to system concepts channel.")
+    (async/go (>! channel {:start-index start-index}))))
+
+(defn index-concepts-by-id
+  "Bulk index the concepts given by the concept-ids"
+  [this context provider-id concept-type concept-ids]
+  (let [channel (:concept-id-channel this)]
+    (info "Adding bulk index request to concept-id channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :concept-type concept-type
+                           :request :index
+                           :concept-ids concept-ids}))))
+
+(defn delete-concepts-from-index-by-id
+  "Bulk delete the concepts given by the concept-ids from the indexes"
+  [this context provider-id concept-type concept-ids]
+  (let [channel (:concept-id-channel this)]
+    (info "Adding bulk delete reqeust to concept-id channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :concept-type concept-type
+                           :request :delete
+                           :concept-ids concept-ids}))))
+
+(defn bootstrap-virtual-products
+  "Initializes virtual products."
+  [this context provider-id entry-title]
+  (let [channel (:virtual-product-channel this)]
+    (info "Adding message to virtual products channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :entry-title entry-title}))))
+
+(defrecord CoreAsyncDispatcher
+  [;; Channel for requesting full provider migration
+   provider-db-channel
+   ;; Channel for requesting single collection/granules migration.
+   ;; Takes maps, e.g., {:collection-id collection-id :provider-id provider-id}
+   collection-db-channel
+   ;; Channel for requesting full provider indexing - collections/granules
+   provider-index-channel
+   ;; Channel for processing collections to index.
+   collection-index-channel
+   ;; Channel for processing data newer than a given date-time.
+   data-index-channel
+   ;; Channel for processing bulk index requests for system concepts (tags, acls, access-groups)
+   system-concept-channel
+   ;; channel for processing bulk index requests by concept-id
+   concept-id-channel
+   ;; channel for bootstrapping virtual products
+   virtual-product-channel])
+
+(def dispatch-behavior
+  "Map of protocol definitions to the implementations of that protocol for the
+  dispatcher."
+  {:migrate-provider migrate-provider
+   :migrate-collection migrate-collection
+   :index-provider index-provider
+   :index-data-later-than-date-time index-data-later-than-date-time
+   :index-collection index-collection
+   :index-system-concepts index-system-concepts
+   :index-concepts-by-id index-concepts-by-id
+   :delete-concepts-from-index-by-id delete-concepts-from-index-by-id
+   :bootstrap-virtual-products bootstrap-virtual-products})
+
+(extend CoreAsyncDispatcher
```

Comment:
You will want to do the `extend`ing in the `core` namespace where the API/protocol is defined. This will let you use that namespace and only that namespace for all implementations.